pkg/controllers/provisioning/scheduling/node.go

```diff
@@ -184,3 +164,39 @@ func (n Node) hasCompatibleResources(resourceList v1.ResourceList, it cloudprovi
 	}
 	return true
 }
+
+func (n Node) hasCompatibleNodeAffinity(affinity *v1.Affinity, it cloudprovider.InstanceType) bool {
+	if affinity == nil || affinity.NodeAffinity == nil || affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil {
+		return true
+	}
+
+	for _, term := range affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
+		if term.MatchExpressions != nil {
+			if !compatibleInstanceType(v1alpha5.NewRequirements(term.MatchExpressions...), it) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func compatibleInstanceType(requirements v1alpha5.Requirements, it cloudprovider.InstanceType) bool {
+	if !requirements.Get(v1.LabelInstanceTypeStable).Has(it.Name()) {
+		return false
+	}
+	if !requirements.Get(v1.LabelArchStable).Has(it.Architecture()) {
+		return false
+	}
+	if requirements.Get(v1.LabelOSStable).Intersection(
```

Comment:
Can't you just do 

```!it.OperatingSystems().HasAny(requirements.Get(v1.LabelOSStable))```

Reply:
I don't think so, the requirement may not specify an OS in which case any are ok.  We'd have to call .Values() on the requirements version which would panic due to calling values on a complement set.  The two different set types here are making the code more ugly than it should be.

Reply:
Another option is to implement HasAny on our specialized set type that encapsulates this. Totally optional.

Reply:
Good call, I didn't want to change much when fixing it for 0.7.2, but I've got the time now to make a bit less ugly.